python/ray/data/datasource/huggingface_datasource.py

```diff
@@ -91,8 +91,13 @@ def get_read_tasks(
         import pyarrow
 
         def _read_dataset(dataset: "datasets.IterableDataset") -> Iterable[Block]:
-            for batch in dataset.with_format("arrow").iter(batch_size=self._batch_size):
-                block = pyarrow.Table.from_pydict(batch)
+            for batch in dataset.iter(batch_size=self._batch_size):
+                # Block could be a pyarrow.Table or a dict, depending on the
+                # formatting configuration of the HF Dataset.
```

Comment:
Nit: If this is always true, could we add an assertion? 
```suggestion
                # Block could be a pyarrow.Table or a dict, depending on the
                # formatting configuration of the HF Dataset.
                assert isinstance(batch, (pyarrow.Table, dict)), f"Expected `pyarrow.Table` or `dict`, but got {type(batch)}"
```

Reply:
Ah, so it looks like the actual root cause of the issue is that `IterableDataset`s do not properly implement `set_format`/`with_format` according to this issue: https://github.com/huggingface/datasets/issues/3444

Looks like there is no current ETA for implementing this method, so in the meantime, I think we need to explicitly choose which batch types we support. According to `with_format` [docs](https://huggingface.co/docs/datasets/v2.14.5/en/package_reference/main_classes#datasets.Dataset.with_format),  the supported formats are: `[None, 'numpy', 'torch', 'tensorflow', 'pandas', 'arrow', 'jax'].` 

Out of these, I propose we support the following: `None` (`dict`), `numpy`, `pandas`, and `arrow`. Since these are data libraries closest to arrow, and the others are more ML framework-specific.

Reply:
Ah, interesting.

Yeah, makes sense to me.